internal/edittree/edittree.go

```diff
@@ -0,0 +1,1165 @@
+// Note: Values pulled from arrays will have the same approximate
+// ordering in the final term.
+func (e *EditTree) Filter(paths []ast.Ref) *ast.Term {
+	if e.value == nil {
+		return nil
+	}
+
+	// Separate out keys for this level.
+	// In the event of paths like "a", "a/b", "a/b/c", the "a" path will win out.
+	// Nil keys, such as "" or [], are not permitted. (legacy behavior)
+	pathMap := make(map[ast.Value][]ast.Ref, len(paths))
+	renderNowList := []*ast.Term{}
+	for i := range paths {
+		path := paths[i]
+		switch {
+		case len(path) == 0:
+			continue // ignore nil paths, such as "" and [].
+		case len(path) == 1:
+			renderNowList = append(renderNowList, path[0])
+		default: // len(path) > 1
+			if _, ok := pathMap[path[0].Value]; !ok {
+				pathMap[path[0].Value] = []ast.Ref{}
+			}
+			pathMap[path[0].Value] = append(pathMap[path[0].Value], path[1:])
+		}
+	}
+	renderNow := ast.NewSet(renderNowList...)
+	// Clear everything out of the pathMap that has a renderNow candidate.
+	for k := range pathMap {
+		if renderNow.Contains(ast.NewTerm(k)) {
+			delete(pathMap, k)
+		}
+	}
+
+	// Now that we've reached the target, we can start rendering everything beneath us in the tree.
+	switch e.value.Value.(type) {
+	case ast.Object:
+		out := make([][2]*ast.Term, 0, renderNow.Len()+len(pathMap))
+		// Render any finished paths.
+		renderNow.Foreach(func(k *ast.Term) {
+			if e.Exists(ast.Ref{k}) {
+				subtreeResult, _ := e.RenderAtPath(ast.Ref{k})
+				out = append(out, [2]*ast.Term{k, subtreeResult})
+			}
+		})
+		// Recursively descend remaining paths.
+		for k, p := range pathMap {
+			if e.Exists(ast.Ref{ast.NewTerm(k)}) {
+				child, _ := e.Unfold(ast.Ref{ast.NewTerm(k)})
+				subtreeResult := child.Filter(p)
+				out = append(out, [2]*ast.Term{ast.NewTerm(k), subtreeResult})
+			}
+		}
+		return ast.ObjectTerm(out...)
+	case ast.Set:
+		out := make([]*ast.Term, 0, renderNow.Len()+len(pathMap))
+		// Render any finished paths.
+		renderNow.Foreach(func(k *ast.Term) {
+			if e.Exists(ast.Ref{k}) {
+				subtreeResult, _ := e.RenderAtPath(ast.Ref{k})
+				out = append(out, subtreeResult)
+			}
+		})
+		// Recursively descend remaining paths.
+		for k, p := range pathMap {
+			if e.Exists(ast.Ref{ast.NewTerm(k)}) {
+				child, _ := e.Unfold(ast.Ref{ast.NewTerm(k)})
+				subtreeResult := child.Filter(p)
+				out = append(out, subtreeResult)
+			}
+		}
+		return ast.SetTerm(out...)
+	case *ast.Array:
+		// No early exit here, because we might have just deletes on the
+		// original array. We build a new Array with modified/deleted keys.
+		out := make([]*ast.Term, 0, renderNow.Len()+len(pathMap))
+		// Sort array indexes before descending.
+		idxList := make([]*ast.Term, 0, len(pathMap))
+		renderNow.Foreach(func(k *ast.Term) {
+			idxList = append(idxList, k)
+		})
+		for k := range pathMap {
+			idxList = append(idxList, ast.NewTerm(k))
+		}
+		sort.Sort(termSlice(idxList))
```

Comment:
This sorting operation is required so that the we preserve the *relative ordering* between the Terms coming from this Array.

Example:

 - Source Term: `{"a": ["a", "b", "c", "d", "e", "f"]}`
 - `Filter` paths: `"a/5", "a/0", "a/2"`
 - Result Term: `{"a": ["a", "c", "f"]}`